File: backend/app/services/scheduler.py
"""Планировщик интервальных повторений (CLAUDE.md §3.6).

Алгоритм спрятан за протоколом `ReviewScheduler`: заменить SM-2 на FSRS можно
подменой реализации, вызывающий код не меняется.
"""

from __future__ import annotations

from dataclasses import dataclass, replace
from datetime import datetime, timedelta
from typing import Final, Protocol

MIN_EASINESS_FACTOR: Final = 1.3
DEFAULT_EASINESS_FACTOR: Final = 2.5

# Порог качества: ниже — повторение сбрасывается.
FAILURE_THRESHOLD: Final = 3

FIRST_INTERVAL_DAYS: Final = 1
SECOND_INTERVAL_DAYS: Final = 6

MIN_QUALITY: Final = 0
MAX_QUALITY: Final = 5


@dataclass(frozen=True, slots=True)
class ReviewSnapshot:
    """Снимок состояния повторения. Неизменяемый: планировщик возвращает новый."""

    easiness_factor: float = DEFAULT_EASINESS_FACTOR
    repetitions: int = 0
    interval_days: int = 0
    due_at: datetime | None = None

    @property
    def is_new(self) -> bool:
        return self.repetitions == 0


class ReviewScheduler(Protocol):
    """Контракт планировщика: из состояния и оценки получить новое состояние."""

    def review(self, state: ReviewSnapshot, quality: int, now: datetime) -> ReviewSnapshot: ...


class SM2Scheduler:
    """Классический SM-2."""

    def review(self, state: ReviewSnapshot, quality: int, now: datetime) -> ReviewSnapshot:
        if not MIN_QUALITY <= quality <= MAX_QUALITY:
            raise ValueError(f"качество ответа {quality} вне диапазона {MIN_QUALITY}–{MAX_QUALITY}")

        easiness_factor = self._next_easiness_factor(state.easiness_factor, quality)

        if quality < FAILURE_THRESHOLD:
            # Провал сбрасывает цепочку: вопрос вернётся завтра.
            repetitions = 0
            interval_days = FIRST_INTERVAL_DAYS
        else:
            repetitions = state.repetitions + 1
            interval_days = self._next_interval(repetitions, state.interval_days, easiness_factor)

        return replace(
            state,
            easiness_factor=easiness_factor,
            repetitions=repetitions,
            interval_days=interval_days,
            due_at=now + timedelta(days=interval_days),
        )

    @staticmethod
    def _next_easiness_factor(easiness_factor: float, quality: int) -> float:
        delta = 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)
        return max(MIN_EASINESS_FACTOR, easiness_factor + delta)

    @staticmethod
    def _next_interval(repetitions: int, interval_days: int, easiness_factor: float) -> int:
        if repetitions == 1:
            return FIRST_INTERVAL_DAYS
        if repetitions == 2:
            return SECOND_INTERVAL_DAYS
        return max(FIRST_INTERVAL_DAYS, round(interval_days * easiness_factor))


def get_scheduler() -> ReviewScheduler:
    """Точка подмены алгоритма для всего приложения."""
    return SM2Scheduler()

File: mobile/lib/domain/scheduler.dart
/// Планировщик интервальных повторений (CLAUDE.md §3.6).
///
/// Порт серверной `app/services/scheduler.py`. Алгоритм спрятан за
/// интерфейсом: заменить SM-2 на FSRS можно подменой реализации, вызывающий
/// код при этом не меняется.
library;

const double kMinEasinessFactor = 1.3;
const double kDefaultEasinessFactor = 2.5;

/// Порог качества: ниже — цепочка повторений сбрасывается.
const int kFailureThreshold = 3;

const int kFirstIntervalDays = 1;
const int kSecondIntervalDays = 6;

const int kMinQuality = 0;
const int kMaxQuality = 5;

/// Снимок состояния повторения. Неизменяемый: планировщик возвращает новый.
class ReviewSnapshot {
  const ReviewSnapshot({
    this.easinessFactor = kDefaultEasinessFactor,
    this.repetitions = 0,
    this.intervalDays = 0,
    this.dueAt,
  });

  final double easinessFactor;
  final int repetitions;
  final int intervalDays;
  final DateTime? dueAt;

  bool get isNew => repetitions == 0;

  ReviewSnapshot copyWith({
    double? easinessFactor,
    int? repetitions,
    int? intervalDays,
    DateTime? dueAt,
  }) =>
      ReviewSnapshot(
        easinessFactor: easinessFactor ?? this.easinessFactor,
        repetitions: repetitions ?? this.repetitions,
        intervalDays: intervalDays ?? this.intervalDays,
        dueAt: dueAt ?? this.dueAt,
      );
}

/// Контракт планировщика: из состояния и оценки получить новое состояние.
abstract interface class ReviewScheduler {
  ReviewSnapshot review(ReviewSnapshot state, int quality, DateTime now);
}

/// Классический SM-2.
class Sm2Scheduler implements ReviewScheduler {
  const Sm2Scheduler();

  @override
  ReviewSnapshot review(ReviewSnapshot state, int quality, DateTime now) {
    if (quality < kMinQuality || quality > kMaxQuality) {
      throw ArgumentError(
        'качество ответа $quality вне диапазона $kMinQuality–$kMaxQuality',
      );
    }

    final double easiness = _nextEasinessFactor(state.easinessFactor, quality);

    final int repetitions;
    final int intervalDays;
    if (quality < kFailureThreshold) {
      // Провал сбрасывает цепочку: вопрос вернётся завтра.
      repetitions = 0;
      intervalDays = kFirstIntervalDays;
    } else {
      repetitions = state.repetitions + 1;
      intervalDays = _nextInterval(repetitions, state.intervalDays, easiness);
    }

    return ReviewSnapshot(
      easinessFactor: easiness,
      repetitions: repetitions,
      intervalDays: intervalDays,
      dueAt: now.add(Duration(days: intervalDays)),
    );
  }

  static double _nextEasinessFactor(double easiness, int quality) {
    final double delta = 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02);
    final double next = easiness + delta;
    return next < kMinEasinessFactor ? kMinEasinessFactor : next;
  }

  static int _nextInterval(int repetitions, int intervalDays, double easiness) {
    if (repetitions == 1) {
      return kFirstIntervalDays;
    }
    if (repetitions == 2) {
      return kSecondIntervalDays;
    }
    final int next = (intervalDays * easiness).round();
    return next < kFirstIntervalDays ? kFirstIntervalDays : next;
  }
}

/// Точка подмены алгоритма для всего приложения.
ReviewScheduler getScheduler() => const Sm2Scheduler();
